### PR TITLE
rebase: `move_commits`: do not remove parents of target commits which are outside the target set

### PR DESCRIPTION
**Pull request by bnjmnt4n**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [X] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/commands/rebase.rs [hunk 1]
```diff
@@ -873,17 +874,33 @@ fn move_commits(
...
+                            new_parents.push(parent_id.clone());
+                        } else if let Some(parents) =
+                                connected_target_commits_internal_parents.get(parent_id) {
+                            new_parents.extend(parents.iter().cloned());
+                        } else if !new_children.iter().any(|new_child| {
+                                mut_repo.index().is_ancestor(new_child.id(), parent_id) }) {
+                            new_parents.push(parent_id.clone());
+                        }
+                    }
+                    new_parents
```
Comment (yuja) [inline cli/src/commands/rebase.rs:903]: It seems a bit odd that the target root is special cased (even if it was a merge.)

For example, rebasing `d::e` off `a|b` removes `a|b` from `d`, but not from `e`. Isn't it better to exclude all ancestors of the target roots?

```
    create_commit(&test_env, &repo_path, "a", &[]);
    create_commit(&test_env, &repo_path, "b", &[]);
    create_commit(&test_env, &repo_path, "c", &[]);
    create_commit(&test_env, &repo_path, "d", &["a", "b"]);
    create_commit(&test_env, &repo_path, "e", &["d", "a", "b"]);
    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
    @      e
    ├─┬─╮
    ◉ │ │  d
    ╰─┬─╮
      │ ◉  b
      ◉ │  a
      ├─╯
    ◉ │  c
    ├─╯
    ◉
    "###);

    test_env.jj_cmd_ok(&repo_path, &["rebase", "-r", "d::e", "-d", "c"]);
    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
    @      e
    ├─┬─╮
    │ │ ◉  b
    │ ◉ │  a
    │ ├─╯
    ◉ │  d
    ◉ │  c
    ├─╯
    ◉
    "###);
```

Comment (bnjmnt4n) [reply cli/src/commands/rebase.rs:903]: Yeah, I can see where you're coming from, it might also makes sense to treat all the revisions passed as an independent set and move it directly to the destination whilst removing any links to other commits. Here's the thread in Discord where I discussed this: https://discord.com/channels/968932220549103686/1234553998703460423.

The example described shows that with the current behavior, rebasing `-r s::` would differ from `-s s`:

```
❯ jj l
@    vwysomoy benjamin@dev.ofcr.se 1 second ago 5331ba13
├─╮  (empty) d
○ │  kurkunwm benjamin@dev.ofcr.se 30 seconds ago 8d3b09e0
│ │  (empty) c
○ │  svotuppn benjamin@dev.ofcr.se 34 seconds ago 001cebf5
├─╯  (empty) b
○  ksnsnuxr benjamin@dev.ofcr.se 43 seconds ago e55b9272
│  (empty) a
◆  zzzzzzzz root() 00000000

❯ jj rebase -r s:: -d ks
Skipping rebase of commits:
  svotuppn 001cebf5 (empty) b
  kurkunwm 8d3b09e0 (empty) c
Rebased 1 commits onto destination
Working copy now at: vwysomoy ac20feeb (empty) d
Parent commit      : kurkunwm 8d3b09e0 (empty) c

❯ jj l
@  vwysomoy benjamin@dev.ofcr.se 1 second ago ac20feeb
│  (empty) d
○  kurkunwm benjamin@dev.ofcr.se 1 minute ago 8d3b09e0
│  (empty) c
○  svotuppn benjamin@dev.ofcr.se 1 minute ago 001cebf5
│  (empty) b
○  ksnsnuxr benjamin@dev.ofcr.se 1 minute ago e55b9272
│  (empty) a
◆  zzzzzzzz root() 00000000

❯ jj undo
Working copy now at: vwysomoy 5331ba13 (empty) d
Parent commit      : kurkunwm 8d3b09e0 (empty) c
Parent commit      : ksnsnuxr e55b9272 (empty) a

❯ jj rebase -s s -d ks
Skipping rebase of commit svotuppn 001cebf5 (empty) b

❯ jj l
@    vwysomoy benjamin@dev.ofcr.se 1 minute ago 5331ba13
├─╮  (empty) d
○ │  kurkunwm benjamin@dev.ofcr.se 1 minute ago 8d3b09e0
│ │  (empty) c
○ │  svotuppn benjamin@dev.ofcr.se 1 minute ago 001cebf5
├─╯  (empty) b
○  ksnsnuxr benjamin@dev.ofcr.se 1 minute ago e55b9272
│  (empty) a
◆  zzzzzzzz root() 00000000
```

I thought that might be a little inconsistent, but as @martinvonz pointed out we can also explicitly make the behavior different for `-r` and `-s`/`-b`.

Any other opinions on the desired behavior for this?

Comment (jonathantanmy) [reply cli/src/commands/rebase.rs:903]: To @yuja 's comment, I think of it in terms of a more realistic example:

```
@    f
├─╮
│ ○  e
│ ○  d
○ │  c
○ │  b
├─╯
○  a
◆
```

There's a main branch `b` and `c` and a feature branch `d` and `e` that got merged as `f`. If we were to rebase `c` and `f` to some unrelated part of the commit graph, I would expect `f` to still have `a` as an ancestor (through `e` and `d`).

I don't think the logic should change if `a` were instead a direct parent of `f` (that is, `e` and `d` are missing).

Regarding @bnjmnt4n 's question, I did not read the Discord thread, but think that `-r s::` should have the same behavior to `-s s` if possible.

File: cli/tests/test_rebase_command.rs [hunk 2]
```diff
@@ -589,10 +592,11 @@ fn test_rebase_multiple_revisions() {
     │ │ ◉  g
     ╭─┬─╯
     │ ◉  e
-    ◉ │  c
     │ │ ◉  h
     │ │ ◉  f
+    ╭───┤
     │ │ ◉  d
+    ◉ │ │  c
```
Comment (jonathantanmy) [inline cli/tests/test_rebase_command.rs:599]: Thanks - these new behaviors look reasonable and are in line with my mental model of how this command should behave. (I haven't looked at the `test_rebase_revisions_after` ones yet - I'll get back once I've wrapped my mind around how the "before" and "after" features should behave.)

File: cli/src/commands/rebase.rs [hunk 3]
```diff
@@ -665,32 +665,33 @@ fn move_commits(
             .commits(mut_repo.store())
             .try_collect()?;
 
-    // Commits in the target set should only have other commits in the set as
-    // parents, except the roots of the set, which persist their original
-    // parents.
+    // Compute the parents of all commits in the connected target set,
+    // allowing only commits in the target set as parents.
     // If a commit in the set has a parent which is not in the set, but has
     // an ancestor which is in the set, then the commit will have that ancestor
```
[the remark this answers is not in the document]
Comment (bnjmnt4n) [reply cli/src/commands/rebase.rs:671]: A better description is "If a commit in the target set has a parent which is not in the target set but has an ancestor in the target set, the parent's closest ancestors which are in the target set will be added to the commit's parents."

In the case of your example, `f` would have parents `d` (direct parent in the target set), as well as `b` and `c` (`e`'s parents which are in the target set).

[the remark this answers is not in the document]
Comment (jonathantanmy) [reply cli/src/commands/rebase.rs:671]: "Closest ancestors" is not well-defined either. Let me add a commit `z` (not a target commit) before `f`:

```
@    f
○    z
├─╮
│ ○    e
│ ├─╮
│ │ ○  c
○ │ │  d
├─╯ │
○   │  b
├───╯
○  a
◆
```

So a commit in the target set `f` has a parent that is not in the target set `z`. What are the parent's (that is, `z`'s) closest ancestors? It could arguably be (1) `d b c` (as you describe) or (2) `d c` (`b` is excluded because `d` is indisputably closer than `b`).

After some thought, the best way I could describe both would be:
(1) any ancestor of `f` that is a target commit and whose ancestry path (from `f`) does not go through another target commit
(2) the heads of the (the intersection of the ancestors of `f` and the target set)

Of the two, I prefer (2) because it reuses terminology from "Revset language", which means that it's easier to understand and test. (Putting myself in the user's shoes, this is a really contrived example, so I think either is reasonable, as long as we are consistent.) But I think this means rewriting the code (as far as I can tell, the code both before and after this PR implements (1)). I can't think of an efficient algorithm to do (2) off the top of my head, unfortunately. (Git does something similar during fetch negotiation, but the output is one big set of "bottom" commits; here, we need one set of "bottom" commits per target commit, so we cannot directly use Git's algorithm.)

File: cli/tests/test_rebase_command.rs (hunk 2 above)
Comment (jonathantanmy) [reply cli/tests/test_rebase_command.rs:599]: To close the loop on this, the `test_rebase_revisions_after` changes look reasonable. Thanks for including new test cases that test what happens when a target commit's parent turns out to be a descendant of a new child - having a specific rule for this case seems a bit inelegant, but I can't think of anything better.

File: cli/src/commands/rebase.rs (hunk 1 above)
Comment (yuja) [reply cli/src/commands/rebase.rs:903]: > The example described shows that with the current behavior, rebasing `-r s::` would differ from `-s s`:

I see. Even though it's inconsistent, I agree that making `-r s::` identical to `-s s` is practically useful.
Perhaps, that's because `-r` specifies nodes to rebase, whereas `-s` specifies edges to break. We need to deduce edges from `-r` sets, so the ambiguity is unavoidable.

Can you include the reason behind this change in the commit message?

File: cli/src/commands/rebase.rs [hunk 4]
```diff
@@ -873,17 +874,33 @@ fn move_commits(
...
+                }
+                // Otherwise:
+                // 1. Keep parents which are within the target set.
+                // 2. Replace parents which are outside the target set but are part of the
+                //    connected target set with their ancestor commits which are in the target
+                //    set.
+                // 3. Keep other parents outside the target set if they are not descendants of the
+                //    new children of the target set.
+                else {
+                    let mut new_parents = vec![];
```
Comment (yuja) [inline cli/src/commands/rebase.rs:891]: Perhaps, `new_parents` can be precomputed in a way that skipped ancestor edges are also preserved?

```
    create_commit(&test_env, &repo_path, "a", &[]);
    create_commit(&test_env, &repo_path, "b", &["a"]);
    create_commit(&test_env, &repo_path, "c", &["a"]);
    create_commit(&test_env, &repo_path, "d", &["a"]);
    create_commit(&test_env, &repo_path, "e", &["c", "b"]);
    create_commit(&test_env, &repo_path, "f", &["e"]);
    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
    @  f
    ◉    e
    ├─╮
    │ ◉  b
    ◉ │  c
    ├─╯
    │ ◉  d
    ├─╯
    ◉  a
    ◉
    "###);

    // TODO: edge f->e->b is lost
    test_env.jj_cmd_ok(&repo_path, &["rebase", "-r", "c|f", "-d", "d"]);
    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
    @  f
    ◉  c
    ◉  d
    │ ◉  e
    ╭─┤
    │ ◉  b
    ├─╯
    ◉  a
    ◉
    "###);
```

File: cli/src/commands/rebase.rs [hunk 5]
```diff
@@ -873,17 +874,33 @@ fn move_commits(
...
+                    for parent_id in commit.parent_ids() {
+                        if target_commit_ids.contains(parent_id) {
+                            new_parents.push(parent_id.clone());
+                        } else if let Some(parents) =
+                                connected_target_commits_internal_parents.get(parent_id) {
+                            new_parents.extend(parents.iter().cloned());
+                        } else if !new_children.iter().any(|new_child| {
```
Comment (yuja) [inline cli/src/commands/rebase.rs:898]: I think this should be an error, but I'm not sure if we handle other loop cases differently.

File: cli/src/commands/rebase.rs
Comment (yuja) [inline cli/src/commands/rebase.rs:1]: Maybe this can be addressed later, but is it possible to move `move_commits()` to jj-lib? The current `test_rebase_command.rs` is hard to track because CLI tests are slow so we have to reuse the environment.

File: cli/src/commands/rebase.rs (hunk 4 above)
Comment (martinvonz) [reply cli/src/commands/rebase.rs:891]: Interesting case. I don't think I would actually expect the dependency on `b` to be preserved.

File: cli/src/commands/rebase.rs (hunk 5 above)
Comment (martinvonz) [reply cli/src/commands/rebase.rs:898]: We have this code for handling loops already: https://github.com/martinvonz/jj/blob/19563fee7483ecd6b5f56200520ac81c84ea16c5/cli/src/commands/rebase.rs#L713-L743

I'm not sure it's worth having that code because it also seems like strange behavior to support but I don't feel strongly. My impression was that @bnjmnt4n also didn't feel strongly either way.

File: cli/src/commands/rebase.rs (hunk 4 above)
Comment (bnjmnt4n) [reply cli/src/commands/rebase.rs:891]: My thought process for the implementation is that when selecting the target commits to rebase, we only want to preserve ancestry information for those commits, not any other commits. However, I'm happy to change the implementation if a consensus is reached. My personal use of this feature has largely been limited to rebasing linear stacks of commits to before/after other locations, so maybe others with different usecases might want to make suggestions as well.

File: cli/src/commands/rebase.rs (hunk 5 above)
Comment (bnjmnt4n) [reply cli/src/commands/rebase.rs:898]: > I think this should be an error, but I'm not sure if we handle other loop cases differently.

In this case, I feel like the user's intent in calling rebase `--before`/`--after` is to move the commits to the new location, making whatever changes to the commit graph are necessary, so this was acceptable to me. `--before`/`--after` allows you to move the root of the target set before its original parents, so I thought moving the other commits of the target set before any of its original parents is acceptable.

File: cli/src/commands/rebase.rs (hunk 4 above)
Comment (yuja) [reply cli/src/commands/rebase.rs:891]: Hmm, in my mental model, `rebase -r c|f` is roughly equivalent to `rebase -r c::f` and `abandon e'`. So I expect the `f->b` edge is preserved so long as `rebase -r c::f` works that way.

Comment (yuja) [reply cli/src/commands/rebase.rs]: Just commenting why I think the `f->e->b` edge should be preserved:
1. disconnected `-r` targets are rebased as if they were all connected
2. in order to achieve that, internal edges across skipped nodes are preserved
3. so it seems better to apply the same rule to external edges

In practice, I wouldn't try this kind of weird rebase, and I wouldn't mind if `f->e->b` was removed. My point is that it's easier to reason about the behavior if we have a consistent model (such as `rebase -r c|f` is equivalent to `rebase --keep -r c::f` + `abandon c|f` + `abandon e'`.)

Comment (bnjmnt4n) [reply cli/src/commands/rebase.rs]: I guess making the logic more consistent is also a good idea. I will update the PR to change this.

Comment (martinvonz) [reply cli/src/commands/rebase.rs]: I think the other reasonable behavior is to drop this PR and let `-s x` and `-r x::` be different. I think I would actually prefer that. That seems even more consistent. Then we don't even have to worry about treating the roots of the connected set differently. It's weird to me if `rebase -r c|f -d a` preserves the edge but `rebase -r f -d a` does not. It doesn't seem that `f` should depend on `b` any more because we're also rebasing `c`. Thoughts?

Comment (martinvonz) [reply cli/src/commands/rebase.rs]: Yet another option is to make `-s` not preserve external edges. I copied it from Mercurial but I'm not sure how useful that behavior is.

Comment (yuja) [reply cli/src/commands/rebase.rs]: > I think the other reasonable behavior is to drop this PR and let `-s x` and `-r x::` be different.

That sounds also good to me.

> It's weird to me if `rebase -r c|f -d a` preserves the edge but `rebase -r f -d a` does not. It doesn't seem that `f` should depend on `b` any more because we're also rebasing `c`.

Since `rebase -r c|f -d a` is not identical to `rebase -r c -d a && rebase -r f -d a`, I won't be surprised if `-r c|f` preserve the `f->e->b` edge, but `-r f` doesn't.

> Yet another option is to make `-s` not preserve external edges. I copied it from Mercurial but I'm not sure how useful that behavior is.

I occasionally do `rebase -s` one branch whose descendants are merged with the other branches. In this case, the current `rebase -s` behavior is convenient.

Comment (bnjmnt4n) [reply cli/src/commands/rebase.rs]: Based on this discussion, should I close this PR? I can open a separate one which moves the `move_commits` function into the `rewrite` module.

Also, is https://github.com/martinvonz/jj/pull/3607 still desired? (I was working initially working on this PR to make the behavior of `rebase -r/-s` consistent before migrating `rebase -s` to the `move_commits` function to add support for `--before/after`.)

Comment (martinvonz) [reply cli/src/commands/rebase.rs]: Yes, my inclination is to drop this PR and let `jj rebase -s foo` and `jj rebase -r foo::` behave differently. Sorry to have wasted your time :(

> Also, is #3607 still desired?

I think it's enough to support `-A/-B` with `-r`, and it might get complicated to support it with `-s` and its different behavior for merges into the set.

Comment (bnjmnt4n) [reply cli/src/commands/rebase.rs]: Apart from `--insert-before/after`, I was thinking of `rebase --keep` as well. I was planning to migrate the existing `rebase -s/-b` to use `move_commits`, then from there modify `move_commits` to allow keeping the original target commits as well. WDYT about that? (I don't think think it will be very complicated to add a flag to persist external merge parents for `-s`/`-b`.)

Comment (martinvonz) [reply cli/src/commands/rebase.rs]: That all sounds good to me. Thanks!

Comment (martinvonz) [reply cli/src/commands/rebase.rs]: #3788 made me think about this thread again. Similar to `-s`, it's a bit ugly that we can't get the same behavior with `-r`. Maybe we should add a `--preserve-external-edges` to preserve edges coming into revisions that are not the roots of the set. I think we could then replace the current `-s foo` implementation by `-r foo:: --preserve-external-edges` then, and the current `-b foo -d bar` by `-r '(bar..foo)::' --preserve-external-edges`.

Comment (yuja) [reply cli/src/commands/rebase.rs]: Do we need `--preserve-external-edges` flag? iirc, there are some oddity in `--preserve-external-edges -rREVS` behavior (especially when `REVS` aren't contiguous), but would it matter in practice?

I think it's be a bit confusing if `-s`/`-b` implies `--preserve-external-edges`, but `-r` doesn't. If the "preserve" behavior is practically good, it can be turned on by default.

**Comment by bnjmnt4n** [comment]
Given @martinvonz's comment on [Discord](https://discord.com/channels/968932220549103686/968932220549103689/1269634348504846346), should I re-open this issue?

**Comment by martinvonz** [comment]
Yes, I suspect that's best (for context: what I said on Discord is that I suspect I was wrong when I said `-s X` and `-r X::` should behave differently). I'm not sure how much it matters in practice for cases like the one we discussed above because they seem rare, and I'm not sure what users expect in those cases. However, just the fact that it removes the difference between `-s X` and `-r X::` seems like a strong reason to do it. Sorry to have wasted your time.

**Review by martinvonz — APPROVED** [review]
LGTM. Thanks again!

File: lib/src/rewrite.rs [hunk 6]
```diff
@@ -710,17 +710,33 @@ pub fn move_commits(
...
+                        if target_commit_ids.contains(parent_id) {
+                            new_parents.push(parent_id.clone());
+                        } else if let Some(parents) =
+                                connected_target_commits_internal_parents.get(parent_id) {
+                            new_parents.extend(parents.iter().cloned());
+                        } else if !new_children.iter().any(|new_child| {
+                                mut_repo.index().is_ancestor(new_child.id(), parent_id) }) {
+                            new_parents.push(parent_id.clone());
+                        }
+                    }
```
Comment (yuja) [inline lib/src/rewrite.rs:738]: Any thoughts on https://github.com/martinvonz/jj/pull/3605#discussion_r1587018019 ?

I don't think it would matter in practice, but I think it's nice that both internal and external edges are processed somewhat consistently. Only the roots are detached.

This is not a blocker, btw.

File: cli/tests/test_rebase_command.rs
Comment (yuja) [inline cli/tests/test_rebase_command.rs:1]: nit: maybe better to include parents in log output. it's hard to follow the graph edges.
```rust
fn get_log_output(test_env: &TestEnvironment, repo_path: &Path) -> String {
    let template = "separate(': ', branches, parents.map(|c| c.branches()))";
    test_env.jj_cmd_success(repo_path, &["log", "-T", template])
}
```

File: lib/src/rewrite.rs (hunk 6 above)
Comment (bnjmnt4n) [reply lib/src/rewrite.rs:738]: I will spend some time (probably next week) to look at this.
